### PR TITLE
CA-217301: Block SXM when SR doesn't have VDI_MIRROR capability

### PR DESCRIPTION
**Pull request by sharady**
Signed-off-by: Sharad Yadav <sharad.yadav@citrix.com>

### CODE REVIEW

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 1]
```diff
@@ -93,31 +93,51 @@ open Storage_interface
...
 open Fun
 
-let assert_sr_support_migration ~__context ~vdi_map ~remote =
+let assert_sr_support_operations ~__context ~vdi_map ~remote ~ops ~_type =
+  let op_supported_on_source_sr vdi ops =
+    (* Check VDIs must not be present on SR which doesn't have required capability *)
+    let source_sr = Db.VDI.get_SR ~__context ~self:vdi in
+    let sr_record = Db.SR.get_record_internal ~__context ~self:source_sr in
+    let sr_features = Xapi_sr_operations.features_of_sr ~__context sr_record in
+    List.iter (fun op ->
```
Comment (lindig) [inline ocaml/xapi/xapi_vm_migrate.ml]: This could be also implemented with `if not (List.for_all Smint.(has_capability op sr_features) ops) then raise … else ()`.  I think it would be a little clearer because it expresses the expectation that all `ops` must meet a requirement, otherwise an exception is thrown. There is a similar situation below.

**Comment by lindig** [comment]
Could this check have been implemented in `xapi_vm_lifecycle.vm`?

**Comment by sharady** [comment]
Since checks for `vif_map` and `vdi_map` already exists in `assert_can_migrate`. It seems to be a better place to keep this check in same function. I guess XC also uses `assert_can_migrate` call to allow/disallow vm-migration. Though we can move this function to any helpers.

File: ocaml/xapi/xapi_vm_migrate.ml (hunk 1 above)
Comment (sharady) [inline ocaml/xapi/xapi_vm_migrate.ml]: Thats a good suggestion Thanks :)
Updating PR.

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 2]
```diff
@@ -93,31 +93,47 @@ open Storage_interface
...
-        | _ -> []
-      in
-      if not (List.exists (fun cp -> cp = Smint.(string_of_capability Vdi_snapshot)) sm_capabilities) then
-        raise (Api_errors.Server_error(Api_errors.sr_does_not_support_migration, [Ref.string_of sr]))
-    ) vdi_map
+  begin match _type with
+  | `intra_pool ->
+    (* Don't fail if all VDIs going to reside on same SR *)
+    let vdi_map' = List.filter (fun (vdi,sr) -> Db.VDI.get_SR ~__context ~self:vdi <> sr) vdi_map in
+    if vdi_map' <> [] then
```
Comment (lindig) [inline ocaml/xapi/xapi_vm_migrate.ml]: You don't need to test for the empty list. List.iter will simply do nothing when the list is empty. I'd write it like this: `List.filter ... |> List.iter ...` but keeping the `let` is fine, too.

**Comment by sharady** [comment]
Thanks @lindig, Updated the same.

**Comment by lindig** [comment]
This looks fine to me but would like @robhoes or @euanh to take a look, too.

File: ocaml/idl/datamodel.ml [hunk 3]
```diff
@@ -959,7 +959,7 @@ let _ =
   error Api_errors.too_many_storage_migrates [ "number" ]
     ~doc:"You reached the maximal number of concurrently migrating VMs." ();
   error Api_errors.sr_does_not_support_migration [ "sr" ]
-    ~doc:"You attempted to migrate a VDI on SR which doesn't have snapshot capability" ();
+    ~doc:"You attempted to migrate a VDI to or from SR which doesn't support migration" ();
```
Comment (euanh) [inline ocaml/idl/datamodel.ml]: to or from *an* SR

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 4]
```diff
@@ -93,31 +93,46 @@ open Storage_interface
...
-        match List.filter (fun (_, r) -> r.API.sM_type = sr_type) sm_record with
-        | [ _, plugin ] -> plugin.API.sM_capabilities
-        | _ -> []
-      in
-      if not (List.exists (fun cp -> cp = Smint.(string_of_capability Vdi_snapshot)) sm_capabilities) then
-        raise (Api_errors.Server_error(Api_errors.sr_does_not_support_migration, [Ref.string_of sr]))
-    ) vdi_map
+  begin match _type with
+  | `intra_pool ->
+    (* Don't fail if all VDIs going to reside on same SR *)
```
Comment (robhoes) [inline ocaml/xapi/xapi_vm_migrate.ml]: Really _all_ VDIs on the _same_ SR?

So before this patch, we checked for `Smint.Vdi_snapshot`, but only in the cross-pool case. Now we check for `Smint.Vdi_snapshot` as well as `Smint.Vdi_mirror` in both the cross-pool and intra-pool case, but in the intra-pool case only if the storage is actually migrated. Seems like this addresses multiple issues at once? If this is as intended, the reasons need to be explained in the commit message.

**Comment by sharady** [comment]
Thanks @robhoes, I have update the commit message, please have a look.

**Comment by robhoes** [comment]
@sharady Thanks for the explanation. This being the case, I think that we do not need to special case cross-pool: filtering SRs that are the same on the source and destination will be a no-op. We can remove the `match _type with` and keep only the intra-pool case to simplify the function. We don't even need the `~_type` parameter.

**Comment by sharady** [comment]
@robhoes thanks for the review. I have removed the unwanted `match`
